### ISSUE TITLE
Try to go back to publishing directly to Maven Central

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#454]
It appears that the sonatype folks are willing to help us. So if we reconfigure our build for sonatype, they may provide us with a faster access...

So we need to go through the song and dance and retool our build setup accordingly.

https://central.sonatype.org/pages/gradle.html

https://central.sonatype.org/pages/ossrh-guide.html

cc @richardstartin  @marshallpierce

**Comment by blacelle** [comment]
@lemire I deploy regularly another jar into sonatype, with more luck than you observed with RoaringBitmap, but with other (non-blockng but painful) issues related to encryption keys. Hence I could also give a hand into this.

**Comment by richardstartin** [comment]
I'm no longer actively using the library so if @blacelle has time that would be great

**Comment by brianf** [comment]
Not sure if this reduces the pain, but I came across this today: https://github.com/saket/startship

**Comment by lemire** [comment]
@blacelle Thanks for the offer. I think that the next step would be a pull request that would modify our setup to go directly to sonatype. It would be great if someone could get such a pull request started so that we can at least discuss it.

This setup should be comparable to what we have right now:

1. We want to remain with our gradle build. It is massively faster than anything we had with Apache Maven and it works reliably.
2. We want to run tests only once when releasing, not twice as we used to do with Apache Maven. That's a big deal because tests can take several minutes so we don't want to double that. I do not want us to go back to running tests twice for no good reason. I understand that we want to run tests during a release process, but I do not understand why we would run them twice.
3. We want a reliable one liner, that is, releasing should be on command line that does everything, like we have currently. In the old Apache Maven setup, I would have to stand there for 10 minutes, while tests run, then I would have to enter a passphrase at a given moment or face a timeout and failed release. Furthermore, people on twitter are quite clear that you do not need to go to the sonatype web GUI to close, wait 5 minutes, then release... that it is all trivially automated (as with bintray currently).
4. It should be portable, evidently. That is, it should run on any operating system. If needed, let us use docker so that we know precisely what the setup is during the release. I never want to go back to having to install specific packages on Linux to get the release to go through. There is really no good reason that releasing Java software should require me to tweak your operating system.


For reference, the way it works right now, with bintray, thanks to @marshallpierce 's help, is that it is a single command line that does the whole release process. It is entirely automated as it should be. Automation during releases is important... I cannot stress this enough... you do not want the release process to depend on one guy doing everything right... you want reliable, trustable processes and for that, it needs to be automated, no manual.

Currently, getting into bintray is that easy, and then, in theory, getting through sonatype only requires pressing a button, you can watch me pressing the button there:

https://www.youtube.com/watch?v=5EuASOahecA

It sometimes work, but, most times, it appears to fail. It is not clear whether we can tell why it fails (see video).

>  I deploy regularly another jar into sonatype, with more luck than you observed with RoaringBitmap

Look at the video. I am sure that your process is one command like, it takes a few seconds and bang, you are on Maven Central. People tell me that we are doing it wrong. I am really eager to learn what we are doing wrong.

**Comment by lemire** [comment]
To be precise, people on twitter are very clear that this song and dance (which is not needed on bintray) described in the sonatype documentation...

https://central.sonatype.org/pages/releasing-the-deployment.html

... is antiquated and not at all necessary.

In my video, when releasing from bintray, I only watch... it all does its thing automatically... 

We do not want to go back to this manual process... especially given that the twitter crowd specifically tells us that it is ancient history.

**Comment by brianf** [comment]
We are working on the docs to make it clear that using the UI is just one way to do the promotion. The reality is that probably 80% or more of the releases coming through our systems are totally headless. There are countless examples on the web showing how people have done it. We have a blog in flight that is helping to consolidate more coverage for those ways. I agree with you that pushing buttons is antiquated, but it's useful for some people that do things in frequently and just want a quick win.

**Comment by brianf** [comment]
Also for the video coming from Bintray, it should be known that they built that integration without ever working with us, so it was reverse engineered and not even using the latest mechanisms. So it's not surprising that it was unstable and there wasn't a lot we could do about it because we didn't have the ability to change how it worked.

**Comment by blacelle** [comment]
@brianf Could you provide some fresh+solid source/reference implemementation to re-introduce RoaringBitmap releases to sonatype servers with Gradle? They are multiple people with goodwill here, but I'm sure none of us want to get back to Sonatype hell we used to encounter.

**Comment by brianf** [comment]
I'm not a Gradle guy but these links provide several sensible ways to automate: https://chris.banes.dev/publishing-to-maven-central/
https://jeroenmols.com/blog/2021/02/04/migratingjcenter/
https://proandroiddev.com/publishing-your-first-android-library-to-mavencentral-be2c51330b88

**Comment by lemire** [comment]
Thanks @brianf for the pointers.

**Comment by brianf** [comment]
https://github.com/rwinch/gradle-publish-ossrh-sample

**Comment by blacelle** [comment]
Started looking into it. I'll early open a draft PR, just to show the progress.

**Comment by blacelle** [comment]
> https://proandroiddev.com/publishing-your-first-android-library-to-mavencentral-be2c51330b88

Relies on cliking buttons.

> https://github.com/rwinch/gradle-publish-ossrh-sample

looks solid and (ultra) fresh.

https://github.com/RoaringBitmap/RoaringBitmap/pull/456

**Comment by blacelle** [comment]
Navigating into dark waters.

Anybody with Gradle+Kotlin proficiency to give a hand here ?

The following code won't compile in build.gradle.kts while OK in build.gradle. 

https://github.com/rwinch/gradle-publish-ossrh-sample#configure-nexus-publishing-plugin
```
nexusPublishing {
	repositories {
		sonatype {
			if (project.hasProperty("sonatypeUsername")) {
				username = project.sonatypeUsername
			}
			if (project.hasProperty("sonatypePassword")) {
				password = project.sonatypePassword
			}
		}
	}
}
```

https://github.com/RoaringBitmap/RoaringBitmap/pull/456/commits/87fafdc27c39f0bb69db62c0ffd7fb76c75a217e#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R241

**Comment by blacelle** [comment]
Got it: https://github.com/marcphilipp/nexus-publish-plugin#kotlin-dsl

**Comment by blacelle** [comment]
@lemire Could you indicate the current release/deploy procedure?

**Comment by lemire** [comment]
```
./gradlew release
```

with bintrayUser set to a bintray key.

The script asks for the version number one wants to release and then, finally, it may ask for the next version.

Then to post to Maven Central, I am supposed to go online and press a button, see the video. If not for the final step (going online and trying to sync to maven), it is all fast, reliable and automated. I simply cannot get it wrong... except, I guess, if I entered the wrong version number.

**Comment by blacelle** [comment]
(I fully agree regarding the need for a dummy-proof release procedure)

**Comment by marshallpierce** [comment]
In case it's helpful, I've got publication to the Sonatype repo working in a multiproject build like this one in https://bitbucket.org/marshallpierce/kotlin-new-relic/src/master/build.gradle.kts. 

It doesn't help with the slowness and unreliability of Sonatype, but here's the procedure:

1. `./gradlew release` -- this creates and populates a "staging repo". This assumes you've already done the GPG key setup needed to sign artifacts, which is a pain.
2. Go to https://oss.sonatype.org/#stagingRepositories and "close" the repo. This will validate the various things maven central wants like GPG signatures. This sometimes takes a really long time, so I don't have it done automatically in the build since it would time out occasionally and fail the release process.
3. Once the "close" is done, "release" the closed repo. This also sometimes takes a really long time and/or fails for no reason.

**Comment by lemire** [comment]
@marshallpierce That's a helpful comment.

**Comment by lemire** [comment]
@marshallpierce Having to go to the sonatype web GUI to close the artefact is *not* a major problem. It only becomes a problem when it is slow and when it fails. 

Anyhow, as far as the current PR is concerned, we do not need to have it fully automated. It should only upload the artefacts. That would be good enough.

**Comment by marshallpierce** [comment]
OK, I'll take a stab at it since I'm already in the middle of doing the same for various other libraries. What a great use of time this is for all the OSS maintainers affected by the Bintray shutdown :roll_eyes:

**Comment by brianf** [comment]
If you want to move over to the new system S01,  Just create a new issue here to request migration: issues.sonatype.org/projects/OSSRH

It will only change the url you use to access the system and it will be much faster and more stable.

**Comment by lemire** [comment]
@brianf Thanks Brian.

**Comment by lemire** [comment]
@marshallpierce I have released what I think is a valid version by effectively doing

` ./gradlew  -Pversion=0.9.10 publishToSonatype `

... it takes about 4 mins. Then I go to oss.sonatype.org and I close... it takes a few minutes, then I hit release and it seems to work.

However, it does not update anything in git, so the snapshot version in `gradle.properties` is unchanged and there is no git tag. I made these changes manually... is that what you expect to happen?

Then I tried to do

` ./gradlew publishToSonatype `

And it *looked* like it worked but it really did not:

```
$  ./gradlew    publishToSonatype

BUILD SUCCESSFUL in 17s
19 actionable tasks: 16 executed, 3 up-to-date
```

But it was too fast and it did nothing.

So I am trying to document the process:

1. ` ./gradlew  -Pversion=0.9.10 publishToSonatype `
2. update `gradle.properties`, commit
3. create a git release (tag), commit
4. go to oss.sonatype.org and close
5. wait a few minutes and go back to oss.sonatype.org and release

Is that it?

**Comment by lemire** [comment]
It does look like it worked...
https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/0.9.10/

**Comment by marshallpierce** [comment]
That will work, but the existing `./gradlew release` flow should work as well because `afterReleaseBuild` depends on `publishToSonatype` for each of the published subprojects. That is:

1. `./gradlew release`, follow prompts for updated version etc as usual
2. go to oss.sonatype.org and close
3. wait a few minutes and go back to oss.sonatype.org and release

The `-Pversion=foo` stuff was just a way to test manually that it could upload artifacts that would pass the `close` process.

**Comment by lemire** [comment]
Ok. I am going to close this issue for now.

I think that @blacelle worked on additional features, but what we have is enough to close the issue.

**Comment by lemire** [comment]
I have created a sonatype issue at https://issues.sonatype.org/browse/OSSRH-67401 following @brianf 's comment.

**Comment by lemire** [comment]
@brianf We switched to the new system. It was so fast that I thought that my test had failed, but there it was in the GUI.

It is now on par with systems like npm.

**Comment by brianf** [comment]
Love it. I'm glad we could get that done quickly for you.

**Comment by lemire** [comment]
@marshallpierce


> ./gradlew release, follow prompts for updated version etc as usual
> go to oss.sonatype.org and close
> wait a few minutes and go back to oss.sonatype.org and release

If I do as you suggest, then nothing appears at sonatype and I am not sure that something gets uploaded.

https://github.com/RoaringBitmap/RoaringBitmap/issues/497

**Comment by lemire** [comment]
Ah. One needs to go to https://s01.oss.sonatype.org/ 
Not https://oss.sonatype.org/ 

These two sites only appear identical.

**Comment by marshallpierce** [comment]
Yep, that's correct -- it's not clear to me why they didn't just make the existing site faster instead of recreating it and causing the bifurcation. Oh well. I've only just now started using `s01` on one project and so far it seems to work.

**Comment by brianf** [comment]
The old system is on hugely scaled infra but the levels just exceeded what
could be thrown at it without underlying software architecture changes. We
are revamping the whole thing, but in the interim, the easiest and
quickest, low risk was to shard the systems essentially.

**Comment by marshallpierce** [comment]
Gotcha. Well, we're glad to have s01 either way; it's a lot nicer to have fast feedback!
